File: windows/resourcestore.go
package main

import (
	"encoding/json"
	"os"
	"path/filepath"

	"github.com/google/uuid"
)

// PingResource is a user-visible "is this reachable" tile - separate from
// SavedConfig, purely diagnostic (no server/psk/etc, just a name + URL).
// Reachability itself is checked by the frontend directly via fetch(),
// which goes through the OS's real network stack and is therefore subject
// to whatever the system routing table currently says - a resource that's
// normally blocked and starts responding once the Phantom tunnel connects
// (or one that stays unreachable even connected) is the whole point of the
// feature, so no Go-side ping method is needed here, just storage.
type PingResource struct {
	ID   string `json:"id"`
	Name string `json:"name"`
	URL  string `json:"url"`
}

const resourcesFileName = "resources.json"

var defaultResources = []PingResource{
	{Name: "YouTube", URL: "https://www.youtube.com"},
	{Name: "Discord", URL: "https://discord.com"},
	{Name: "ChatGPT", URL: "https://chatgpt.com"},
	{Name: "Claude", URL: "https://claude.ai"},
}

func loadResources() ([]PingResource, error) {
	dir, err := configDir()
	if err != nil {
		return nil, err
	}
	path := filepath.Join(dir, resourcesFileName)
	data, err := os.ReadFile(path)
	if err != nil {
		if !os.IsNotExist(err) {
			return nil, err
		}
		// First run: seed the built-in defaults (with real IDs) so the panel
		// isn't empty and the user can delete/customize from there on.
		seeded := make([]PingResource, len(defaultResources))
		for i, r := range defaultResources {
			seeded[i] = PingResource{ID: uuid.NewString(), Name: r.Name, URL: r.URL}
		}
		if err := saveResources(dir, seeded); err != nil {
			return nil, err
		}
		return seeded, nil
	}
	var resources []PingResource
	if err := json.Unmarshal(data, &resources); err != nil {
		return nil, err
	}
	return resources, nil
}

func saveResources(dir string, resources []PingResource) error {
	data, err := json.Marshal(resources)
	if err != nil {
		return err
	}
	return os.WriteFile(filepath.Join(dir, resourcesFileName), data, 0600)
}

func addResource(name, url string) (PingResource, error) {
	dir, err := configDir()
	if err != nil {
		return PingResource{}, err
	}
	resources, err := loadResources()
	if err != nil {
		return PingResource{}, err
	}
	r := PingResource{ID: uuid.NewString(), Name: name, URL: url}
	if err := saveResources(dir, append(resources, r)); err != nil {
		return PingResource{}, err
	}
	return r, nil
}

func deleteResource(id string) error {
	dir, err := configDir()
	if err != nil {
		return err
	}
	resources, err := loadResources()
	if err != nil {
		return err
	}
	filtered := resources[:0]
	for _, r := range resources {
		if r.ID != id {
			filtered = append(filtered, r)
		}
	}
	return saveResources(dir, filtered)
}

File: android/app/src/main/java/com/phantom/vpn/ResourceStore.kt
package com.phantom.vpn

import android.content.Context
import android.content.SharedPreferences
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey
import org.json.JSONArray
import org.json.JSONObject
import java.util.UUID

/**
 * One "is this reachable" tile - separate from SavedConfig, purely diagnostic (just a
 * name + URL). Reachability is checked directly against the URL from the app's own
 * process, which goes through Android's normal networking - once the VPN tunnel is up,
 * VpnService routes this app's own traffic through it too (same as every other app on
 * the device), so a normally-blocked resource starting to respond is the whole point.
 */
data class PingResource(val id: String, val name: String, val url: String)

/** Mirrors ConfigStore's storage pattern exactly, just for the smaller resource-tile model. */
object ResourceStore {
    private const val RESOURCES_KEY = "ping_resources" // JSON array: [{"id":..,"name":..,"url":..}]

    private val defaults = listOf(
        "YouTube" to "https://www.youtube.com",
        "Discord" to "https://discord.com",
        "ChatGPT" to "https://chatgpt.com",
        "Claude" to "https://claude.ai",
    )

    @Volatile
    private var cached: SharedPreferences? = null

    @Synchronized
    private fun prefs(context: Context): SharedPreferences {
        cached?.let { return it }
        val opened = try {
            val masterKey = MasterKey.Builder(context)
                .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
                .build()
            EncryptedSharedPreferences.create(
                context, "phantom_secure_prefs", masterKey,
                EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
                EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM
            )
        } catch (t: Throwable) {
            FileLog.e("EncryptedSharedPreferences init failed, falling back to plain prefs", t)
            null
        } ?: context.getSharedPreferences("phantom_plain_prefs", Context.MODE_PRIVATE)
        cached = opened
        return opened
    }

    fun loadAll(context: Context): List<PingResource> {
        val p = prefs(context)
        val raw = p.getString(RESOURCES_KEY, null)
        if (raw == null) {
            // First run: seed the built-in defaults so the page isn't empty and the
            // user can delete/customize from there on.
            val seeded = defaults.map { (name, url) -> PingResource(UUID.randomUUID().toString(), name, url) }
            saveAll(context, seeded)
            return seeded
        }
        return runCatching {
            val arr = JSONArray(raw)
            (0 until arr.length()).map { i ->
                val obj = arr.getJSONObject(i)
                PingResource(obj.getString("id"), obj.getString("name"), obj.getString("url"))
            }
        }.getOrDefault(emptyList())
    }

    fun saveAll(context: Context, resources: List<PingResource>) {
        val arr = JSONArray()
        resources.forEach { r ->
            arr.put(JSONObject().apply { put("id", r.id); put("name", r.name); put("url", r.url) })
        }
        prefs(context).edit().putString(RESOURCES_KEY, arr.toString()).apply()
    }

    fun add(context: Context, name: String, url: String): PingResource {
        val r = PingResource(id = UUID.randomUUID().toString(), name = name, url = url)
        saveAll(context, loadAll(context) + r)
        return r
    }

    fun delete(context: Context, id: String) {
        saveAll(context, loadAll(context).filterNot { it.id == id })
    }
}
